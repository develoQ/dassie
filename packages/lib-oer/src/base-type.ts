import { ParseError, SerializeError } from "./utils/errors"
import type { ParseContext, SerializeContext } from "./utils/parse"
import { TagClass, TagMarker, tagClassMarkerMap } from "./utils/tag"

export type AnyOerType = OerType<unknown>

export type Infer<TOerType extends AnyOerType> = TOerType extends OerType<
  infer T,
  unknown
>
  ? T
  : never

export type InferSerialize<TOerType extends AnyOerType> =
  TOerType extends OerType<unknown, infer T> ? T : never

export interface ParseOptions {
  allowNoncanonical?: boolean
}

export type IntermediateSerializationResult = readonly [
  serialize: (
    context: SerializeContext,
    offset: number
    // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
  ) => SerializeError | void,
  length: number
]

export abstract class OerType<TParseValue, TSerializeValue = TParseValue> {
  _tag: readonly [tagValue: number, tagClass: TagMarker] | undefined
  _optional = false

  abstract parseWithContext(
    context: ParseContext,
    offset: number
  ): readonly [value: TParseValue, length: number] | ParseError
  abstract serializeWithContext(
    value: TSerializeValue
  ): IntermediateSerializationResult | SerializeError

  parse(
    input: Uint8Array,
    offset = 0,
    options: ParseOptions = {}
  ):
    | { success: true; value: TParseValue; length: number }
    | { success: false; failure: ParseError } {
    const context: ParseContext = {
      uint8Array: input,
      dataView: new DataView(input.buffer, input.byteOffset, input.byteLength),
      allowNoncanonical: options.allowNoncanonical ?? false,
    }
    const result = this.parseWithContext(context, offset)

    if (result instanceof ParseError) {
      return { success: false, failure: result }
    }

    if (!context.allowNoncanonical && result[1] !== input.byteLength - offset) {
      return {
        success: false,
        failure: new ParseError(
          "non-canonical encoding - additional bytes present after the expected end of data",
          input,
          result[1]
        ),
      }
    }

    return { success: true, value: result[0], length: result[1] }
  }

  serialize(
    value: TSerializeValue
  ):
    | { success: true; value: Uint8Array }
    | { success: false; failure: SerializeError } {
    const prediction = this.serializeWithContext(value)

    if (prediction instanceof SerializeError) {
      return { success: false, failure: prediction }
    }

    const [serialize, length] = prediction

    const uint8Array = new Uint8Array(length)

    const result = serialize(
      {
        uint8Array,
        dataView: new DataView(
          uint8Array.buffer,
          uint8Array.byteOffset,
          uint8Array.byteLength
        ),
      },
      0
    )

    return result instanceof SerializeError
      ? { success: false, failure: result }
      : { success: true, value: uint8Array }
  }

  tag(
    tagValue: number,
    tagClass: Exclude<TagClass, "universal"> = "context"
  ): this {
    this._tag = [tagValue, tagClassMarkerMap[tagClass]] as const
    return this
  }

  optional(): this {
    this._optional = true
    return this
  }
}
