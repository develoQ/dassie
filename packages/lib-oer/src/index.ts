export type { Infer, OerType, AnyOerType, ParseOptions } from "./base-type"
export * from "./bitstring"
export * from "./boolean"
export * from "./choice"
export * from "./enumerated"
export * from "./integer-as-bigint"
export * from "./integer-as-number"
export * from "./octet-string"
export * from "./sequence-of"
export * from "./sequence"
export * from "./string"
