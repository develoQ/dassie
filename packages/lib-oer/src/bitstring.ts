import type { ReadonlyTuple } from "type-fest"

import { OerType } from "./base-type"
import { ParseError, SerializeError } from "./utils/errors"
import {
  parseLengthPrefix,
  predictLengthPrefixLength,
  serializeLengthPrefix,
} from "./utils/length-prefix"
import type { BaseContext, ParseContext, SerializeContext } from "./utils/parse"

type InferBitstringValue<T extends number> = ReadonlyTuple<boolean, T>

interface BitstringOptions {
  variableLength?: boolean
}

export const parseBitstring = (
  context: ParseContext,
  offset: number,
  size: number
) => {
  const byteLength = Math.ceil(size / 8)
  const output = Array.from({ length: size })
  for (let byteIndex = 0; byteIndex < byteLength; byteIndex++) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const byte = context.uint8Array[offset + byteIndex]!

    for (let bitIndex = 0; bitIndex < 8; bitIndex++) {
      if (byteIndex * 8 + bitIndex >= size) {
        break
      }
      const bit = (byte >> (7 - bitIndex)) & 1
      output[byteIndex * 8 + bitIndex] = bit === 1
    }
  }
  return output
}

export const serializeBitstring = (
  context: BaseContext,
  offset: number,
  input: InferBitstringValue<number>
) => {
  const { uint8Array } = context
  const inputAsArray = input as boolean[]
  const byteLength = Math.ceil(inputAsArray.length / 8)

  for (let byteIndex = 0; byteIndex < byteLength; byteIndex++) {
    let byte = 0
    for (let bitIndex = 0; bitIndex < 8; bitIndex++) {
      const index = byteIndex * 8 + bitIndex
      if (index >= inputAsArray.length) break
      byte |= inputAsArray[index] ? 1 << (7 - bitIndex) : 0
    }
    uint8Array[offset + byteIndex] = byte
  }
}

export const bitstring = <TBitDefinition extends number>(
  bits: TBitDefinition,
  { variableLength = false }: BitstringOptions = {}
) => {
  const OerVariableBitstring = class extends OerType<
    InferBitstringValue<TBitDefinition>
  > {
    parseWithContext(context: ParseContext, offset: number) {
      const { uint8Array } = context
      const result = parseLengthPrefix(context, offset)

      if (result instanceof ParseError) {
        return result
      }

      const [length, lengthOfLength] = result

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const unusedBits = uint8Array[offset + lengthOfLength]!

      if (unusedBits > 7) {
        return new ParseError(
          "unable to read bitstring - unused bits greater than 7",
          uint8Array,
          offset + 1
        )
      }

      const output = parseBitstring(
        context,
        offset + lengthOfLength + 1,
        (length - 1) * 8 - unusedBits
      )

      return [
        output as InferBitstringValue<TBitDefinition>,
        lengthOfLength + length,
      ] as const
    }

    serializeWithContext(input: ReadonlyTuple<boolean, TBitDefinition>) {
      const byteLength = Math.ceil(bits / 8)
      const unusedBits = 8 - (bits % 8)
      const length = byteLength + 1
      const lengthOfLength = predictLengthPrefixLength(length)
      if (lengthOfLength instanceof SerializeError) {
        return lengthOfLength
      }

      return [
        (context: SerializeContext, offset: number) => {
          const { uint8Array } = context
          const lengthOfLength = serializeLengthPrefix(
            length,
            uint8Array,
            offset
          )

          if (lengthOfLength instanceof SerializeError) {
            return lengthOfLength
          }

          uint8Array[offset + lengthOfLength] = unusedBits
          serializeBitstring(context, offset + lengthOfLength + 1, input)
          return
        },
        lengthOfLength + length,
      ] as const
    }
  }

  const OerFixedBitstring = class extends OerType<
    InferBitstringValue<TBitDefinition>
  > {
    parseWithContext(context: ParseContext, offset: number) {
      const byteLength = Math.ceil(bits / 8)

      if (context.uint8Array.length - offset < byteLength) {
        return new ParseError(
          "unable to read bitstring value - end of buffer",
          context.uint8Array,
          offset
        )
      }

      const output = parseBitstring(context, offset, bits)

      return [
        output as InferBitstringValue<TBitDefinition>,
        byteLength,
      ] as const
    }

    serializeWithContext(input: InferBitstringValue<TBitDefinition>) {
      const inputAsArray = input as boolean[]
      if (inputAsArray.length !== bits) {
        return new SerializeError(
          `unable to serialize bitstring - expected ${bits} bits, got ${inputAsArray.length}`
        )
      }

      return [
        (context: SerializeContext, offset: number) => {
          serializeBitstring(context, offset, input)
        },
        Math.ceil(bits / 8),
      ] as const
    }
  }
  return variableLength ? new OerVariableBitstring() : new OerFixedBitstring()
}
