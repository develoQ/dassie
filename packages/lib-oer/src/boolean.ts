import { OerType } from "./base-type"
import { ParseError } from "./utils/errors"
import type { ParseContext, SerializeContext } from "./utils/parse"

export const boolean = () => {
  const OerBoolean = class extends OerType<boolean> {
    parseWithContext(
      { uint8Array, allowNoncanonical }: ParseContext,
      offset: number
    ) {
      const value = uint8Array[offset]
      if (typeof value === "undefined") {
        return new ParseError(
          "unable to read boolean value - end of buffer",
          uint8Array,
          offset
        )
      }
      if (!allowNoncanonical && value !== 0 && value !== 0xff) {
        return new ParseError(
          "unable to read boolean value - non-canonical",
          uint8Array,
          offset
        )
      }
      return [!!value, 1] as const
    }
    serializeWithContext(input: boolean) {
      return [
        ({ uint8Array }: SerializeContext, offset: number) => {
          uint8Array[offset] = input ? 0xff : 0
        },
        1,
      ] as const
    }
  }
  return new OerBoolean()
}
