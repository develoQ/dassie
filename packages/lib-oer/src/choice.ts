import { AnyOerType, Infer, InferSerialize, OerType } from "./base-type"
import { ParseError, SerializeError } from "./utils/errors"
import type { ParseContext, SerializeContext } from "./utils/parse"
import {
  TAG_MARKER_CONTEXT,
  TagMarker,
  parseTag,
  predictTagLength,
  serializeTag,
  tagMarkerClassMap,
} from "./utils/tag"

/**
 * Change all properties of an interface to type never
 */
type Never<T> = { [P in keyof T]?: never }

/**
 * Returns a union of interfaces where each member is an object containing a single property which is of such type as inferred from the corresponding member of the input interface.
 */
type InferOptionValues<T extends Record<string, AnyOerType>> = {
  [K in keyof T]-?: { [P in K]: Infer<T[P]> } & Never<
    Pick<T, Exclude<keyof T, K>>
  >
}[keyof T]

type InferOptionSerializeValues<T extends Record<string, AnyOerType>> = {
  [K in keyof T]-?: { [P in K]: InferSerialize<T[P]> } & Never<
    Pick<T, Exclude<keyof T, K>>
  >
}[keyof T]

// Because JavaScript's bitwise operators are 32 bits and we are using two bits for the tag class, we only have 30 bits available to encode the tag value
// This should be plenty though since tag values are usually much smaller.
const MAX_SAFE_TAG_VALUE = Number.MAX_SAFE_INTEGER >>> 2

export const choice = <TOptions extends Record<string, AnyOerType>>(
  options: TOptions
) => {
  // Automatic tagging is enabled if none of the values are explicitly tagged
  const isAutomaticallyTagged = Object.values(options).every(
    (oer) => oer._tag == undefined
  )
  let index = 0
  // Tags are encoded here as an integer with the class marker in the lowest two bits and the tag value shifted two bits to the right.
  const keyToTagMap = new Map<keyof TOptions, number>()
  const tagToKeyMap = new Map<number, keyof TOptions>()

  for (const [key, oer] of Object.entries(options)) {
    const tag = isAutomaticallyTagged
      ? (index++ << 2) | TAG_MARKER_CONTEXT
      : oer._tag != undefined
      ? (oer._tag[0] << 2) | oer._tag[1]
      : undefined

    if (tag == undefined) {
      throw new Error(
        `OER choice: ${key} is not tagged and automatic tagging is disabled`
      )
    }

    if (tag > MAX_SAFE_TAG_VALUE) {
      throw new Error(
        `OER choice: ${key} has a tag value of ${tag} which is too large`
      )
    }

    if (tagToKeyMap.has(tag)) {
      throw new Error(
        `OER choice: ${key} has the same tag value as ${String(
          tagToKeyMap.get(tag)
        )}`
      )
    }

    keyToTagMap.set(key, tag)
    tagToKeyMap.set(tag, key)
  }

  Object.fromEntries(
    Object.entries(options).map(([key, oer]) => {
      return [
        key,
        isAutomaticallyTagged
          ? ([index++, TAG_MARKER_CONTEXT] as const)
          : oer._tag,
      ]
    })
  )

  const setHint = Object.entries(keyToTagMap)
    .map(([key, tag]) => `${key}(${String(tag)})`)
    .join(",")

  const OerChoice = class extends OerType<
    InferOptionValues<TOptions>,
    InferOptionSerializeValues<TOptions>
  > {
    parseWithContext(context: ParseContext, offset: number) {
      const tagResult = parseTag(context, offset)

      if (tagResult instanceof ParseError) {
        return tagResult
      }

      const [tag, tagClass, tagLength] = tagResult //?

      if (tag > MAX_SAFE_TAG_VALUE) {
        return new ParseError(
          `unable to parse choice value - tag value ${tag} is too large`,
          context.uint8Array,
          offset
        )
      }

      const key = tagToKeyMap.get((tag << 2) | tagClass)

      if (key == undefined) {
        return new ParseError(
          `unable to read choice value - tag ${tagMarkerClassMap[
            tagClass
          ].toUpperCase()} ${tag} not in set ${setHint}`,
          context.uint8Array,
          offset
        )
      }

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const oer = options[key]!

      const result = oer.parseWithContext(context, offset + tagLength)

      if (result instanceof ParseError) {
        return result
      }

      return [
        { [key]: result[0] } as InferOptionValues<TOptions>,
        tagLength + result[1],
      ] as const
    }

    serializeWithContext(input: InferOptionSerializeValues<TOptions>) {
      const key = Object.keys(input)[0]

      if (key == undefined) {
        return new SerializeError(
          `unable to serialize choice value - no option selected`
        )
      }

      const tag = keyToTagMap.get(key)

      if (tag == undefined) {
        return new SerializeError(
          `unable to serialize choice value - option ${key} not in set ${setHint}`
        )
      }

      const tagValue = tag >>> 2
      const tagClass = (tag & 3) as TagMarker

      const oer = options[key]

      if (oer == undefined) {
        return new SerializeError(
          `unable to serialize choice value - option ${key} not in set ${setHint}`
        )
      }

      const tagLength = predictTagLength(tagValue)

      if (tagLength instanceof SerializeError) {
        return tagLength
      }

      const valueResult = oer.serializeWithContext(input[key])

      if (valueResult instanceof SerializeError) {
        return valueResult
      }

      return [
        (context: SerializeContext, offset: number) => {
          const { uint8Array } = context

          const tagSerializeResult = serializeTag(
            tagValue,
            tagClass,
            uint8Array,
            offset
          )

          if (tagSerializeResult instanceof SerializeError) {
            return tagSerializeResult
          }

          const serializeResult = valueResult[0](context, offset + tagLength)

          if (serializeResult instanceof SerializeError) {
            return serializeResult
          }

          return
        },
        tagLength + valueResult[1],
      ] as const
    }
  }
  return new OerChoice()
}
