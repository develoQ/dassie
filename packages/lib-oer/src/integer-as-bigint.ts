import { OerType } from "./base-type"
import {
  isUnsignedBigint,
  signedBigintByteLength,
  unsignedBigintByteLength,
} from "./utils/bigint"
import { ParseError, SerializeError } from "./utils/errors"
import {
  parseLengthPrefix,
  predictLengthPrefixLength,
  serializeLengthPrefix,
} from "./utils/length-prefix"
import type { ParseContext, SerializeContext } from "./utils/parse"
import { Range, parseRange } from "./utils/range"

interface IntegerOptions {
  minimumValue?: bigint
  maximumValue?: bigint
}

export interface FixedIntegerOptions extends IntegerOptions {
  minimumValue: bigint
  maximumValue: bigint
}

export const UINT_MIN = 0n
export const UINT8_MAX = 255n
export const UINT16_MAX = 65_535n
export const UINT32_MAX = 4_294_967_295n
export const UINT64_MAX = 18_446_744_073_709_551_615n

export const INT8_MIN = -128n
export const INT8_MAX = 127n
export const INT16_MIN = -32_768n
export const INT16_MAX = 32_767n
export const INT32_MIN = -2_147_483_648n
export const INT32_MAX = 2_147_483_647n
export const INT64_MIN = -9_223_372_036_854_775_808n
export const INT64_MAX = 9_223_372_036_854_775_807n

export abstract class OerFixedIntegerBigint extends OerType<bigint> {
  abstract readonly type: "Uint" | "Int"
  abstract readonly size: 8 | 16 | 32 | 64

  constructor(readonly options: FixedIntegerOptions) {
    super()
  }

  parseWithContext({ uint8Array, dataView }: ParseContext, offset: number) {
    if (offset + this.size / 8 > dataView.byteLength) {
      return new ParseError(
        `unable to read fixed length integer of size ${
          this.size / 8
        } bytes - end of buffer`,
        uint8Array,
        uint8Array.byteLength
      )
    }

    const value =
      this.size === 64
        ? dataView[`getBig${this.type}${this.size}`](offset)
        : BigInt(dataView[`get${this.type}${this.size}`](offset))

    if (value < this.options.minimumValue) {
      return new ParseError(
        `unable to read fixed length integer of size ${
          this.size / 8
        } bytes - value ${value} is less than minimum value ${
          this.options.minimumValue
        }`,
        uint8Array,
        offset
      )
    }

    if (value > this.options.maximumValue) {
      return new ParseError(
        `unable to read fixed length integer of size ${
          this.size / 8
        } bytes - value ${value} is greater than maximum value ${
          this.options.maximumValue
        }`,
        uint8Array,
        offset
      )
    }

    return [value, this.size / 8] as const
  }

  serializeWithContext(value: bigint) {
    const serializer = (context: SerializeContext, offset: number) => {
      if (value < this.options.minimumValue) {
        return new SerializeError(
          `integer must be >= ${this.options.minimumValue}`
        )
      }

      if (value > this.options.maximumValue) {
        return new SerializeError(
          `integer must be <= ${this.options.maximumValue}`
        )
      }

      if (this.size === 64) {
        context.dataView[`setBig${this.type}${this.size}`](offset, value)
      } else {
        context.dataView[`set${this.type}${this.size}`](offset, Number(value))
      }

      return
    }
    serializer.size = this.size / 8
    return serializer
  }
}

export class OerUint8Bigint extends OerFixedIntegerBigint {
  readonly type = "Uint"
  readonly size = 8
}
export class OerUint16Bigint extends OerFixedIntegerBigint {
  readonly type = "Uint"
  readonly size = 16
}
export class OerUint32Bigint extends OerFixedIntegerBigint {
  readonly type = "Uint"
  readonly size = 32
}
export class OerUint64Bigint extends OerFixedIntegerBigint {
  readonly type = "Uint"
  readonly size = 64
}

export class OerInt8Bigint extends OerFixedIntegerBigint {
  readonly type = "Int"
  readonly size = 8
}
export class OerInt16Bigint extends OerFixedIntegerBigint {
  readonly type = "Int"
  readonly size = 16
}
export class OerInt32Bigint extends OerFixedIntegerBigint {
  readonly type = "Int"
  readonly size = 32
}
export class OerInt64Bigint extends OerFixedIntegerBigint {
  readonly type = "Int"
  readonly size = 64
}

export class OerVariableUnsignedInteger extends OerType<bigint> {
  parseWithContext(context: ParseContext, offset: number) {
    const { uint8Array, dataView } = context
    const result = parseLengthPrefix(context, offset)

    if (result instanceof ParseError) {
      return result
    }

    const [length, lengthOfLength] = result

    if (length === 0) {
      return new ParseError(
        `unable to read variable length integer - length must not be 0`,
        uint8Array,
        offset
      )
    }

    offset += lengthOfLength

    let value = 0n
    let index = 0
    while (index < length) {
      if (index + 4 < length) {
        value = (value << 32n) + BigInt(dataView.getUint32(offset + index))
        index += 4
      } else {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        value = (value << 8n) + BigInt(uint8Array[offset + index]!)
        index += 1
      }
    }

    return [value, lengthOfLength + length] as const
  }

  serializeWithContext(input: bigint) {
    if (!isUnsignedBigint(input)) {
      return new SerializeError(
        `expected unsigned bigint, got ${typeof input}: ${input}`
      )
    }

    const length = unsignedBigintByteLength(input)
    const lengthOfLengthPrefix = predictLengthPrefixLength(length)

    if (lengthOfLengthPrefix instanceof SerializeError) {
      return lengthOfLengthPrefix
    }

    const serializer = ({ uint8Array }: SerializeContext, offset: number) => {
      const length = unsignedBigintByteLength(input)
      const lengthOfLengthPrefix = serializeLengthPrefix(
        length,
        uint8Array,
        offset
      )

      if (lengthOfLengthPrefix instanceof SerializeError) {
        return lengthOfLengthPrefix
      }

      let remainder: bigint = input

      for (let index = 0; index < length; index++) {
        const byte = remainder & 0xffn
        uint8Array[offset + lengthOfLengthPrefix + length - index - 1] =
          Number(byte)
        remainder >>= 8n
      }

      return
    }
    serializer.size = lengthOfLengthPrefix + length
    return serializer
  }
}

export class OerVariableSignedInteger extends OerType<bigint> {
  parseWithContext(context: ParseContext, offset: number) {
    const { uint8Array } = context
    const result = parseLengthPrefix(context, offset)

    if (result instanceof ParseError) {
      return result
    }

    const [length, lengthOfLength] = result

    if (length === 0) {
      return new ParseError(
        `unable to read variable length integer - length must not be 0`,
        uint8Array,
        offset
      )
    }

    offset += lengthOfLength

    let value = 0n
    let index = 0
    while (index < length) {
      value = (value << 8n) | BigInt(uint8Array[offset + index]!)
      index += 1
    }

    const twosComplement = 2n ** (BigInt(length) * 8n)
    if (value >= twosComplement >> 1n) {
      value = -(twosComplement - value)
    }

    return [value, lengthOfLength + length] as const
  }

  serializeWithContext(input: bigint) {
    const length = signedBigintByteLength(input)
    const lengthOfLengthPrefix = predictLengthPrefixLength(length)

    if (lengthOfLengthPrefix instanceof SerializeError) {
      return lengthOfLengthPrefix
    }

    const serializer = ({ uint8Array }: SerializeContext, offset: number) => {
      const lengthPrefixSerializeResult = serializeLengthPrefix(
        length,
        uint8Array,
        offset
      )

      if (lengthPrefixSerializeResult instanceof SerializeError) {
        return lengthPrefixSerializeResult
      }

      // Calculate two's complement if input is negative
      if (input < 0n) {
        input += 2n ** BigInt(length * 8)
      }

      let remainder: bigint = input

      for (let index = 0; index < length; index++) {
        const byte = remainder & 0xffn
        uint8Array[offset + lengthOfLengthPrefix + length - index - 1] =
          Number(byte)
        remainder >>= 8n
      }

      return
    }
    serializer.size = lengthOfLengthPrefix + length
    return serializer
  }
}

export const integerAsBigint = (range?: Range<bigint>) => {
  const [minimumValue, maximumValue] = parseRange(range)

  if (minimumValue != undefined && maximumValue != undefined) {
    const fixedOptions = {
      minimumValue,
      maximumValue,
    }

    // Fixed size integer encodings
    if (minimumValue >= UINT_MIN) {
      if (maximumValue <= UINT8_MAX) {
        return new OerUint8Bigint(fixedOptions)
      } else if (maximumValue <= UINT16_MAX) {
        return new OerUint16Bigint(fixedOptions)
      } else if (maximumValue <= UINT32_MAX) {
        return new OerUint32Bigint(fixedOptions)
      } else if (maximumValue <= UINT64_MAX) {
        return new OerUint64Bigint(fixedOptions)
      }
    } else {
      if (minimumValue <= INT8_MIN && maximumValue <= INT8_MAX) {
        return new OerInt8Bigint(fixedOptions)
      } else if (minimumValue <= INT16_MIN && maximumValue <= INT16_MAX) {
        return new OerInt16Bigint(fixedOptions)
      } else if (minimumValue <= INT32_MIN && maximumValue <= INT32_MAX) {
        return new OerInt32Bigint(fixedOptions)
      } else if (minimumValue <= INT64_MIN && maximumValue <= INT64_MAX) {
        return new OerInt64Bigint(fixedOptions)
      }
    }
  }

  // Variable size integer encodings
  return minimumValue != undefined && minimumValue >= UINT_MIN
    ? new OerVariableUnsignedInteger()
    : new OerVariableSignedInteger()
}

export const uint8Bigint = () => integerAsBigint([UINT_MIN, UINT8_MAX])
export const uint16Bigint = () => integerAsBigint([UINT_MIN, UINT16_MAX])
export const uint32Bigint = () => integerAsBigint([UINT_MIN, UINT32_MAX])
export const uint64Bigint = () => integerAsBigint([UINT_MIN, UINT64_MAX])

export const int8Bigint = () => integerAsBigint([INT8_MIN, INT8_MAX])
export const int16Bigint = () => integerAsBigint([INT16_MIN, INT16_MAX])
export const int32Bigint = () => integerAsBigint([INT32_MIN, INT32_MAX])
export const int64Bigint = () => integerAsBigint([INT64_MIN, INT64_MAX])
