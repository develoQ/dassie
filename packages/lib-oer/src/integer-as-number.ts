import { OerType } from "./base-type"
import { ParseError, SerializeError } from "./utils/errors"
import type { ParseContext, SerializeContext } from "./utils/parse"
import { FixedRange, Range, parseRange } from "./utils/range"

export interface IntegerAsNumberOptions {
  range: FixedRange<number>
}

export const UINT_MIN_NUMBER = 0
export const UINT8_MAX_NUMBER = 255
export const UINT16_MAX_NUMBER = 65_535
export const UINT32_MAX_NUMBER = 4_294_967_295

export const INT8_MIN_NUMBER = -128
export const INT8_MAX_NUMBER = 127
export const INT16_MIN_NUMBER = -32_768
export const INT16_MAX_NUMBER = 32_767
export const INT32_MIN_NUMBER = -2_147_483_648
export const INT32_MAX_NUMBER = 2_147_483_647

export const createOerFixedIntegerNumber = (
  size: 8 | 16 | 32,
  type: "Uint" | "Int"
) => {
  const byteSize = size / 8

  return class extends OerType<number> {
    constructor(readonly options: IntegerAsNumberOptions) {
      super()
    }

    parseWithContext({ uint8Array, dataView }: ParseContext, offset: number) {
      if (offset + byteSize > dataView.byteLength) {
        return new ParseError(
          `unable to read fixed length integer of size ${byteSize} bytes - end of buffer`,
          uint8Array,
          uint8Array.byteLength
        )
      }

      const value = dataView[`get${type}${size}`](offset)

      if (value < this.options.range[0]) {
        return new ParseError(
          `unable to read fixed length integer of size ${byteSize} bytes - value ${value} is less than minimum value ${this.options.range[0]}`,
          uint8Array,
          offset
        )
      }

      if (value > this.options.range[1]) {
        return new ParseError(
          `unable to read fixed length integer of size ${byteSize} bytes - value ${value} is greater than maximum value ${this.options.range[1]}`,
          uint8Array,
          offset
        )
      }

      return [value, byteSize] as const
    }

    serializeWithContext(value: number) {
      return [
        (context: SerializeContext, offset: number) => {
          if (value < this.options.range[0]) {
            return new SerializeError(
              `integer must be >= ${this.options.range[0]}`
            )
          }

          if (value > this.options.range[1]) {
            return new SerializeError(
              `integer must be <= ${this.options.range[1]}`
            )
          }

          context.dataView[`set${type}${size}`](offset, value)

          return
        },
        byteSize,
      ] as const
    }
  }
}

export const OerUint8Number = createOerFixedIntegerNumber(8, "Uint")
export const OerUint16Number = createOerFixedIntegerNumber(16, "Uint")
export const OerUint32Number = createOerFixedIntegerNumber(32, "Uint")

export const OerInt8Number = createOerFixedIntegerNumber(8, "Int")
export const OerInt16Number = createOerFixedIntegerNumber(16, "Int")
export const OerInt32Number = createOerFixedIntegerNumber(32, "Int")

export const integerAsNumber = (range: Range<number>) => {
  const [minimumValue, maximumValue] = parseRange(range)

  if (minimumValue == undefined || maximumValue == undefined) {
    throw new Error(
      "When using JavaScript numbers, a minimum and maximum must be provided. If your range is actually unbounded, use integerAsBigint instead."
    )
  }

  const fixedOptions: IntegerAsNumberOptions = {
    range: [minimumValue, maximumValue],
  }

  // Fixed size integer encodings
  if (minimumValue >= UINT_MIN_NUMBER) {
    if (maximumValue <= UINT8_MAX_NUMBER) {
      return new OerUint8Number(fixedOptions)
    } else if (maximumValue <= UINT16_MAX_NUMBER) {
      return new OerUint16Number(fixedOptions)
    } else if (maximumValue <= UINT32_MAX_NUMBER) {
      return new OerUint32Number(fixedOptions)
    }
  } else {
    if (minimumValue <= INT8_MIN_NUMBER && maximumValue <= INT8_MAX_NUMBER) {
      return new OerInt8Number(fixedOptions)
    } else if (
      minimumValue <= INT16_MIN_NUMBER &&
      maximumValue <= INT16_MAX_NUMBER
    ) {
      return new OerInt16Number(fixedOptions)
    } else if (
      minimumValue <= INT32_MIN_NUMBER &&
      maximumValue <= INT32_MAX_NUMBER
    ) {
      return new OerInt32Number(fixedOptions)
    }
  }

  throw new TypeError(
    "integerAsNumber only supports up to 32 bits due to JavaScript's limitations, please use integerAsBigint instead"
  )
}

export const uint8Number = () =>
  integerAsNumber([UINT_MIN_NUMBER, UINT8_MAX_NUMBER])
export const uint16Number = () =>
  integerAsNumber([UINT_MIN_NUMBER, UINT16_MAX_NUMBER])
export const uint32Number = () =>
  integerAsNumber([UINT_MIN_NUMBER, UINT32_MAX_NUMBER])
export const int8Number = () =>
  integerAsNumber([INT8_MIN_NUMBER, INT8_MAX_NUMBER])
export const int16Number = () =>
  integerAsNumber([INT16_MIN_NUMBER, INT16_MAX_NUMBER])
export const int32Number = () =>
  integerAsNumber([INT32_MIN_NUMBER, INT32_MAX_NUMBER])
