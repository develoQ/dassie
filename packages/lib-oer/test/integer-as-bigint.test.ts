import { describe, test } from "vitest"

import { integerAsBigint } from "../src/integer-as-bigint"
import { hexToUint8Array } from "../src/utils/hex"
import { parsedOk, serializedOk } from "./utils/result"

describe("integerAsBigint", () => {
  test("should be a function", ({ expect }) => {
    expect(integerAsBigint).toBeTypeOf("function")
  })

  test("should return an object", ({ expect }) => {
    expect(integerAsBigint()).toBeTypeOf("object")
  })

  test("should parse a zero", ({ expect }) => {
    const value = integerAsBigint().parse(hexToUint8Array("00"))
    expect(value).toEqual(parsedOk(1, 0n))
  })

  test("should serialize a zero", ({ expect }) => {
    const value = integerAsBigint().serialize(0n)
    expect(value).toEqual(serializedOk("00"))
  })

  test("should parse a bigint with one byte", ({ expect }) => {
    const value = integerAsBigint().parse(hexToUint8Array("01 12"))
    expect(value).toEqual(parsedOk(2, 0x12n))
  })

  test("should serialize a bigint with one byte", ({ expect }) => {
    const value = integerAsBigint().serialize(0x12n)
    expect(value).toEqual(serializedOk("01 12"))
  })

  test("should serialize a negative bigint with one byte", ({ expect }) => {
    const value = integerAsBigint().serialize(-0x12n)
    expect(value).toEqual(serializedOk("01 ee"))
  })

  test("should serialize a very large number", ({ expect }) => {
    const testValue = 0x12_34_56_78_9a_bc_de_f0_12_34_56_78_9a_bc_de_f0n

    const value = integerAsBigint().serialize(testValue)
    expect(value).toMatchSnapshot()
  })
})
