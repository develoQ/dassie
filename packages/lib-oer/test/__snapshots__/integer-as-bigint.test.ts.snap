// Vitest Snapshot v1

exports[`integerAsBigint > schema uint8 with range 5..8 > should not parse a number that is too high 1`] = `
{
  "failure": [ParseError: unable to read fixed length integer of size 1 bytes - value 9 is greater than maximum value 8],
  "success": false,
}
`;

exports[`integerAsBigint > schema uint8 with range 5..8 > should not parse a number that is too low 1`] = `
{
  "failure": [ParseError: unable to read fixed length integer of size 1 bytes - value 1 is less than minimum value 5],
  "success": false,
}
`;

exports[`integerAsBigint > schema uint8 with range 5..8 > should not serialize a number that is too high 1`] = `
{
  "failure": [SerializeError: integer must be <= 8],
  "success": false,
}
`;

exports[`integerAsBigint > schema uint8 with range 5..8 > should not serialize a number that is too low 1`] = `
{
  "failure": [SerializeError: integer must be >= 5],
  "success": false,
}
`;

exports[`integerAsBigint > schema variable-length signed integer > should reject a zero-length zero 1`] = `
{
  "failure": [ParseError: unable to read variable length integer - length must not be 0],
  "success": false,
}
`;

exports[`integerAsBigint > schema variable-length signed integer > should serialize a very large number 1`] = `
{
  "success": true,
  "value": Uint8Array [
    16,
    18,
    52,
    86,
    120,
    154,
    188,
    222,
    240,
    18,
    52,
    86,
    120,
    154,
    188,
    222,
    240,
  ],
}
`;

exports[`integerAsBigint > schema variable-length unsigned integer > should refuse to serialize a negative bigint 1`] = `
{
  "failure": [SerializeError: expected unsigned bigint, got bigint: -18],
  "success": false,
}
`;

exports[`integerAsBigint > schema variable-length unsigned integer > should reject a zero-length zero 1`] = `
{
  "failure": [ParseError: unable to read variable length integer - length must not be 0],
  "success": false,
}
`;

exports[`integerAsBigint > schema variable-length unsigned integer > should serialize a very large number 1`] = `
{
  "success": true,
  "value": Uint8Array [
    16,
    18,
    52,
    86,
    120,
    154,
    188,
    222,
    240,
    18,
    52,
    86,
    120,
    154,
    188,
    222,
    240,
  ],
}
`;
